File: gherkin/dotnet/Gherkin/CucumberMessages/Types/FeatureChild.cs
using System.Runtime.Serialization;

namespace Gherkin.CucumberMessages.Types
{
    public class FeatureChild
    {
        [DataMember(Name = "scenario")]
        public Scenario Scenario { get; set; }

        [DataMember(Name = "background")]
        public Background Background { get; set; }

        [DataMember(Name = "rule")]
        public Rule Rule { get; set; }

        public FeatureChild()
        {
        }

        public FeatureChild(Scenario scenario, Background background, Rule rule)
        {
            Scenario = scenario;
            Background = background;
            Rule = rule;
        }
    }
}

File: gherkin/dart/lib/src/ast/FeatureChild.dart
import 'package:gherkin/ast.dart';
import 'package:gherkin/core.dart';

class FeatureChild
    implements /*IHasLocation,*/ INullSafetyObject
{
  static const empty = _InvalidFeatureChild();

  final Background background;

  final Scenario scenario;

  final Rule rule;

  const FeatureChild.fromBackground(this.background)
    : scenario = Scenario.empty
    , rule = Rule.empty;

  const FeatureChild.fromScenario(this.scenario)
    : background = Background.empty
    , rule = Rule.empty;

  const FeatureChild.fromRule(this.rule)
    : background = Background.empty
    , scenario = Scenario.empty;

  @override
  bool get isEmpty => false;

  @override
  bool get isNotEmpty => !isEmpty;
}


/// Convenience implementation of an invalid [FeatureChild] instance.
class _InvalidFeatureChild
    extends FeatureChild
{
  const _InvalidFeatureChild()
      : super.fromBackground(Background.empty);

  @override
  bool get isEmpty => true;
}
